src/scipp/ndimage/__init__.py

```diff
@@ -31,25 +31,27 @@ def function(x: Union[Variable, DataArray], **kwargs) -> Union[Variable, DataArr
     return function
 
 
-def _delta_to_positional(x: Union[Variable, DataArray], dim, index):
-    if isinstance(index, int):
-        return index
+def _delta_to_positional(x: Union[Variable, DataArray], dim, index, dtype):
+    if not isinstance(index, Variable):
+        return dtype(index)
     coord = x.coords[dim]
     if not islinspace(coord, dim).value:
         raise CoordError(
             f"Data points not regularly spaced along {dim}. To ignore this, "
-            "provide a plain value (int or float) instead of a scalar variable. "
-            "Note that this will correspond to plain positional indices/offsets.")
-    return (len(coord) - 1) * (index.to(unit=coord.unit) / (coord[-1] - coord[0])).value
+            f"provide a plain value (convertible to {dtype.__name__}) instead of a "
+            "scalar variable. Note that this will correspond to plain positional "
+            "indices/offsets.")
+    pos = (len(coord) - 1) * (index.to(unit=coord.unit) / (coord[-1] - coord[0])).value
+    return dtype(pos)
 
 
-def _positional_index(x: Union[Variable, DataArray], index, name=None):
-    if isinstance(index, (int, Variable)):
-        return [_delta_to_positional(x, dim, index) for dim in x.dims]
+def _positional_index(x: Union[Variable, DataArray], index, name=None, dtype=int):
+    if not hasattr(index, '__iter__'):
```

Comment:
This check looks strange. What cases does it handle?

Reply:
The code later handles anything that is a mapping. I could just check for `not isinstance(index, dict)` and hope no one tries to use other mappings?

Reply:
What if in the future, `Variable` has an iterator? I think a more concrete check would be safer. It could also be `hasattr(index, 'keys')`. This is how `dict.update` identifies dict-like types.